commons/com.b2international.index/src/com/b2international/index/revision/RevisionBranchRef.java

```diff
@@ -129,29 +128,25 @@ public Expression toRevisedInFilter() {
 	public RevisionBranchRef difference(RevisionBranchRef other) {
 		final TreeSet<RevisionSegment> differenceSegments = newTreeSet();
 		
-		Iterator<RevisionSegment> segmentsIterator = segments.iterator();
-		Iterator<RevisionSegment> otherSegmentsIterator = other.segments.iterator();
+		Map<Long, RevisionSegment> segmentsById = Maps.uniqueIndex(segments, RevisionSegment::branchId);
+		Map<Long, RevisionSegment> otherSegmentsById = Maps.uniqueIndex(other.segments, RevisionSegment::branchId);
 		
-		while (segmentsIterator.hasNext() && otherSegmentsIterator.hasNext()) {
-			RevisionSegment nextSegment = segmentsIterator.next();
-			RevisionSegment otherSegment = otherSegmentsIterator.next();
-			if (nextSegment.branchId() != otherSegment.branchId()) {
-				differenceSegments.add(nextSegment);
-				break;
-			}
-			RevisionSegment difference = nextSegment.difference(otherSegment);
+		// compute segment diff for intersecting branch segments
+		for (Long branchId : Sets.intersection(segmentsById.keySet(), otherSegmentsById.keySet())) {
+			RevisionSegment difference = segmentsById.get(branchId).difference(otherSegmentsById.get(branchId));
 			if (!difference.isEmpty()) {
 				differenceSegments.add(difference);
 			}
 		}
 		
-		while (segmentsIterator.hasNext()) {
-			RevisionSegment nextSegment = segmentsIterator.next();
-			if (!nextSegment.isEmpty()) {
-				differenceSegments.add(nextSegment);
+		// add any remaining segments if they are not empty
+		for (Long branchId : Sets.difference(segmentsById.keySet(), otherSegmentsById.keySet())) {
```

Comment:
Do we need `Sets#symmetricDifference` here (segments on _either side_ that are not present on the other), or just the ones which are in `segmentsById` but not in `otherSegmentsById`?

Reply:
No, this method should always return the diff from the current `this` RevisionBranchRef, similar to the `Sets.difference` method.